Add C preprocessor

LiamPattinson opened the pull request:
Adds the beginnings of a provenance-tracking C-preprocessor.

Takes some steps steps towards https://github.com/PlasmaFAIR/fortitude/issues/405.

I'm raising this as a draft PR as I'll be away for a while and won't have much of a chance to work on it further. I might have some time in early October, but after that I won't be able to give it much thought until late November at the earliest. I'd be happy for others to open PRs into this branch in the meantime.

So far handles the following:

- Fortran-appropriate tokenisation
- Backslash line continuations
- `#define` (object macros only)
- `#undef`
- Some conditional control:
  - `#ifdef`
  - `#ifndef`
  - `#else`
  - `#endif`
- `__LINE__`
- `__FILE__`
- `__DATE__`
- `__TIME__`
- Tracks the provenance of each snippet, e.g. code from within a macro expansion is mapped back to the relevant `#define`.

There's still a lot to add though:

- `#if`, `#elif`
  - `defined()`
  - All associated boolean logic...
- `#include`
  - Cross-platform include paths
- Function macros
  - Stringification
  - Concatenation (need to handle old style `/**/` and new style `##`)
- Provenances should be nestable
- Error reporting currently isn't very helpful

It will also need to be integrated into the rest of the code:

- `fortitude preprocess` mode and relevant options
  - Global defines
  - Per-file defines
  - Include path
  - Per-file include path (?)
- Integration into the linter
- Integration into the LSP/formatter (I'd rather not think about this for now!)

Some stretch goals:

- Ability to set user profiles for things like debug mode or to set environment-specific defines. They could then be activated by calling something like `fortitude preprocess --profile=debug`.
- Ability to read from `compile_commands.json` to set per-file defines and include paths.

---

There's a known bug in `#define`, as it currently saves the text string at the point the macro is defined, but it should actually save a token stream, as macro expansion occurs where the macro is used, not where it's defined:

```c
#define X Y
#define Y 5

X  // expands to '5', not 'Y'
```

LiamPattinson committed:
Add rudimentary c preprocessor tokeniser

LiamPattinson committed:
Clippy fixes

LiamPattinson committed:
Add directives to cpp tokeniser

LiamPattinson committed:
Add beginnings of C preprocessor

LiamPattinson committed:
Add ifdef, ifndef, else, endif to preprocessor

LiamPattinson committed:
Preprocess over logical lines

LiamPattinson committed:
Remove TODO comments, minor clean up

LiamPattinson committed:
fmt update

ZedThree commented on crates/fortitude_preprocessor/src/cpp_tokens.rs:
```diff
@@ -0,0 +1,618 @@
+        // Consume an identifier for the directive name.
+        match self.consume_identifier() {
+            Some(Ok(directive)) => {
+                let directive_kind = match directive.text {
```
I think you can replace this with `#[derive(strum_macros::EnumStr)]` on `CppDirectiveKind`, using `#[strum(default)]` on `Stringification`

LiamPattinson replied:
Done in https://github.com/PlasmaFAIR/fortitude/pull/510/commits/3fe737b2ffebf3cb31dc0d6821d3a894b86eef01 :+1:

ZedThree commented on crates/fortitude_preprocessor/src/cpp_tokens.rs:
```diff
@@ -0,0 +1,618 @@
+    }
+
+    #[test]
+    fn test_identifier_tokenization() {
```
Could do pretty much all of these tests as parameterised tests with `test_case`?

ZedThree commented on crates/fortitude_preprocessor/src/cpp.rs:
```diff
@@ -0,0 +1,852 @@
+    UserDefined,
+    /// A macro defined in the source file.
+    FileDefined {
+        start: usize,
+        end: usize,
```
Is the plan to switch to `ruff_text_size::TextRange`?

ZedThree replied:
Yeah, they're definitely a bit more annoying, but `TextRange` is 1/4 the size of `(usize, usize)`, and if we're going to make a whole bunch of these, it's probably worth it.

But definitely do the simple thing and get it working first!

ZedThree commented on crates/fortitude_preprocessor/src/cpp.rs:
```diff
@@ -0,0 +1,852 @@
+
+/// A snippet of text with provenance information.
+pub struct Snippet {
+    text: String,
```
Is this the transformed text? A possible future optimisation would be to use `Cow<str>` here to avoid unnecessary allocations

LiamPattinson replied:
Yes, the snippets will be the final text of the preprocessed file. I haven't used `Cow<str>` before, but I'd be happy to look into it if it'll save some copies.

LiamPattinson replied:
I tried this with `Cow<'a, str>` and ran into a host of lifetime issues. For now I've left it as a `String`, though I was able to use `Cow<'a, str>` in `LogicalLine`.

ZedThree commented on crates/fortitude_preprocessor/src/cpp.rs:
```diff
@@ -0,0 +1,852 @@
+    inner: Vec<Snippet>,
+}
+
+impl Default for Snippets {
```
Is it not possible to do `#[derive(Default)]` for this?

ZedThree commented on crates/fortitude_preprocessor/src/cpp.rs:
```diff
@@ -0,0 +1,852 @@
+/// logical line.
+pub struct LogicalLine {
+    /// The text of the logical line.
+    text: String,
```
Another place for `Cow<str>`?

ZedThree commented on crates/fortitude_preprocessor/src/cpp.rs:
```diff
@@ -0,0 +1,852 @@
+pub struct LogicalLine {
+    /// The text of the logical line.
+    text: String,
+    /// The byte offsets of each character in the logical line
+    /// relative to the start of the source file.
+    byte_offsets: Vec<usize>,
```
Storing offset of each character individually worries me a little bit. Potential optimisation would be `Vec<TextRange>` storing contiguous chunks?

LiamPattinson replied:
Yes, I was planning to optimise that at some point. I was thinking it would only be necessary to store the discontinuities, but thought I would implement it the easy way to begin.

ZedThree replied:
:+1: Sounds good! Make it work first, then make it fast :)

ZedThree commented:
A couple of implementations in Python worth looking at for inspiration:

- [`fortls`, the Fortran LSP server](https://github.com/fortran-lang/fortls/blob/78ad649ab1a866a40591badb094c8e8849bb15c5/fortls/parsers/internal/parser.py#L2028)
  - Only 300 lines (!) and only a minimal implementation, but useful *enough* I think
- [`pcpp`](https://github.com/ned14/pcpp/tree/master)
  - Much more feature complete, has some interesting additional features
  - Has a much larger test suite

LiamPattinson committed:
Use strum_macros for large CppDirectiveKind match statement

LiamPattinson committed:
Use Cow<'_, str> for LogicalLines

LiamPattinson committed:
Fix object macro expansion rule

Now performs macro expansions at the call site, not in the
definition

LiamPattinson committed:
Add function macros to preprocessor

LiamPattinson commented:
I was trying to get some better tests into the preprocessor, and the ones I could find in the wild seem to make liberal use of `/**/` for concatenation and other tricks , so I thought I'd quickly implement that.

In doing so, I got into a bit of bother trying to figure out how `/**/` interacts with escaped newlines, or how either one of them is supposed to work alongside multiline strings. Turns out, the c-preprocessor doesn't get on well with multiline strings _at all_:

```f90
#define x 12
print *, "hello x &
  & x world!", x
```

After preprocessing:

```f90
print *, "hello x &
  & 12 world!", x
```

It interprets the first `x` on line 3 as an identifier and expands it as a macro, while the second `x` is interpreted as being inside of a string and is left alone, whereas the opposite should be true.

I was then wondering why it didn't interpret the `!` as the start of a comment, and therefore also learned that it doesn't have any special handling for Fortran comments whatsoever:

```f90
#define x 12
! x
! "x
```

becomes:

```f90
! 12
! "x
```

This means I'll have to go back and rethink some of my earlier work, as my tokeniser was written with the expectation that comments and multiline strings were meaningful.

LiamPattinson committed:
Tokenise C comments, fix string and F comment handling

Multiline strings and Fortran comments aren't handled by
the preprocessor, now replicates gfortran behaviour.

Additionally assumes that utf-8 errors won't be raised
while processing, removes the need to handle one
layer of Result types.

cpp.rs is temporarily disabled, and will be updated
for the next commit.

LiamPattinson committed:
Update cpp.rs to match tokeniser update

LiamPattinson committed:
Refactor logical lines

Previous algorithm found to be too inflexible to handle
C-style comments over multiple lines.

- Moved LogicalLine & LogicalLines to separate file.
- Implement merge function that consumes two LogicalLine
  to create a new one.
- Algorithm should be adaptable to handle C-style comments
  that span multiple lines.

Also fixes some comments on the performance of byte offset
tracking and the use of usize throughout.

- Byte offsets use BTreeMap and only store discontinuities.
- Use TextSize everywhere.

LiamPattinson committed:
Implement multiline c comments in LogicalLines

Multiline comments are handled by retaining only the
contents in the last line up to the termination.

LiamPattinson committed:
Handle C-style comments when preprocessing

LiamPattinson committed:
Refine preprocessor test with comments and escaped newlines

LiamPattinson commented:
> Also sounds like a good candidate for a rule? `non-portable-C-comment-in-macro`?

Good idea! I don't think it needs to be within a macro necessarily, as you can do concatenation tricks with them anywhere, and yet `pcpp` will always replace them with a space. The best way to handle it would probably be to find the pattern `/*`, checking that it isn't inside of a string, and iterate until finding `*/` or EOF. I suppose you might also need to be careful not to match `/*` inside other C-comments too.

LiamPattinson committed:
Add preprocess CLI command

LiamPattinson committed:
Clean up preprocessor

- move contents of cpp.rs into lib.rs
- move Definition/Definitions into definitions.rs

LiamPattinson committed:
Remove some #[allow(dead_code)] and stale comments

LiamPattinson committed:
Fix missing argument bug for function macros
